Feature/TA-1-Forget-Password

Ahmed-Abdelghany-11 opened the pull request:
This pull request implements the **Forget Password**, **Email Verification**, and **Reset Password** functionalities, integrating them seamlessly across the **data**, **domain**, and **presentation** layers. These enhancements ensure a smoother password recovery experience for users while improving code maintainability and structure. Additionally, minor optimizations and refactoring were done to enhance performance and maintain consistency across the project.

### Changes Included:

#### 🔹 Data Layer:
- Added API endpoints in `AuthRetrofitClient` for:
  - Sending a forget password request.
  - Verifying the reset code.
  - Resetting the password.
- Implemented **Auth Remote Data Source** to manage authentication-related API calls.
- Created request and response models for:
  - `forget_password`
  - `verify_reset_code`
  - `reset_password`
- Developed the **Auth Repository Implementation** to handle data fetching and processing.

#### 🔹 Domain Layer:
- Created new **use cases** for handling:
  - Forget password requests.
  - Reset code verification.
  - Password reset submission.
- Integrated **repository contracts** to ensure structured data handling.

#### 🔹 Presentation Layer:
- Developed dedicated screens for:
  - **Forget Password**
  - **Email Verification**
  - **Reset Password**
- Enhanced UI elements to provide a **more user-friendly experience**.
- Improved widget performance by using `const` where applicable.

#### 🔹 Localization:
- Updated Arabic & English translation files (`ar.json` & `en.json`) to include **new authentication-related strings**.

## 📌 Pull Request Checklist

### ✅ General
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] The PR title follows the required format: `Feature/TA-1-Forget-Password`.
- [x] The branch name follows the required format: `feature/meaningful-name`, `fix/issue-description`, etc.

### 📂 Code Quality
- [x] All filenames are in `snake_case`.
- [x] All functions and variable names follow Dart naming conventions.
- [x] All new code is formatted with `dart format .`

### 🛠 CI/CD Validation
- [x] Linting passes with `dart analyze`.
- [x] The branch name is validated against the allowed patterns.
- [x] The PR title is validated against the required format.

### 🚀 Testing
- [ ] Unit tests are included for all new features and bug fixes.
- [ ] `flutter test --coverage` has been run and the results reviewed.
- [ ] The project builds successfully without errors.

### 📝 Additional Information
This update enhances the **password recovery process**, ensuring a structured and scalable implementation while improving user experience. Future updates will focus on adding **unit tests** to validate functionality and maintain reliability.

### 🔗 Related JIRA Ticket(s):
- [x] JIRA Issue: `TA-1`

### 📸 Screenshots (if applicable):
<div align="center">

|         Forget Password          |        Email Verification        |        Reset Password         |
|:---------------------:|:----------------------:|:----------------------:|
| <img src="https://github.com/user-attachments/assets/df282d7e-5849-4d30-9e25-566a68e4e77d" width="300" alt="Forget Password"> | <img src="https://github.com/user-attachments/assets/0fd2f489-f7cc-4fc4-bafb-179913b47674" width="300" alt="Email Verification"> | <img src="https://github.com/user-attachments/assets/320ec6ff-6b5a-432c-b2ba-33bb6dabb3d4" width="300" alt="Reset Password"> |

</div>

Ahmed-Abdelghany-11 committed:
feat(auth): implement forget password data layer

This commit adds the implementation for the forget password, verify reset code and reset password flows, including:
- Create ForgetPasswordRequestDto and ForgetPasswordResponseDto models.
- Create VerifyResetCodeRequestDto and VerifyResetCodeResponseDto models.
- Create ResetPasswordRequestDto and ResetPasswordResponseDto models.
- Add new APIs for forget password, verify reset code, and reset password in AuthRetrofitClient.
- Implement AuthRemoteDataSource with the new methods.
- Update DI configuration to inject the new dependencies.

Ahmed-Abdelghany-11 committed:
feat(auth): implement reset password flow with use cases and repository

This commit introduces the reset password flow with the following changes:

-   Created `ForgetPasswordUseCase`, `VerifyResetCodeUseCase`, and `ResetPasswordUseCase` in the domain layer.
-   Implemented `AuthRepoImpl` with `forgetPassword`, `verifyResetCode`, and `resetPassword` methods in the data layer.
-   Defined `AuthRepo` interface in the domain layer.
- Updated dependency injection to register the new use cases and the repository implementation.
- Refactor the file structure.

Ahmed-Abdelghany-11 committed:
feat(forget-password): implement forget password, reset password and email verification cubit

This commit introduces the forget password feature, including:
- Reset password functionality with new cubit `ResetPasswordCubit` and state `ResetPasswordState`.
- Email verification functionality with new cubit `EmailVerificationCubit` and state `EmailVerificationState`.
- Forget password functionality with updated cubit `ForgetPasswordCubit` and state `ForgetPasswordState`.
- Create Forget password request action `ForgetPasswordRequestAction`
- Create Reset password request action `ResetPasswordRequestAction`
- Create Email verification request action `EmailVerificationRequestAction` and `ResendEmailVerificationRequestAction`

Ahmed-Abdelghany-11 committed:
feat(forget_password): implement presentation forget password flow with email verification and password reset

This commit introduces the "forget password" flow, which includes the following:

-   **Dependencies:** Added `pinput` package for OTP input and updated dependencies versions.
-   **New Screens:** Added `EmailVerificationScreen` and `ResetPasswordScreen`.
-   **New Widgets:**
    -   Created `Header` widget for consistent screen headers.
    -   Created `ResendCode` widget for resending verification codes.
    -   Created `ResetPasswordForm` widget for the password reset form.
    -   Created `VerificationCodeInput` widget for the OTP input field.
    -   Created `ForgetPasswordForm` widget for the forget password form.
-   **New Logic:**
    -   Created `EmailVerificationCubit` and `ResetPasswordCubit` for state management.
-   -  add `Validator` class to validate form inputs
-   **Routing:** Added routes for `resetPasswordRoute` and `emailVerificationRoute`.
-   **Translations:** Updated `ar.json` and `en.json` with new translations for the flow.
-  **Dependency Injection**: Add the new cubit to the service locator.
-  **Validation**: Add validation for the inputs forms.

Ahmed-Abdelghany-11 committed:
refactor(auth): Update dependencies and format code

- Update `auth_retrofit_client.g.dart`:
  - Remove unnecessary blank lines and format the code.
  - Simplify the constructor and make it consistent.
  - Remove redundant `copyWith` calls, instead use the direct options.
- Update `en.json`:
  - Fix typo in `InvalidPassword`.
- Update `di.config.dart`:
  - Reformat dependency registration to improve readability.
  - Update dependencies registration order.
  - Use short function notation for simple factory functions.
- Update `locale_keys.g.dart`:
  - Reformat code to follow linting.
- Update `main.dart`:
  - Remove demo home page and `MyHomePage`.
  - Set the initial route to `forgetPasswordRoute`.

Ahmed-Abdelghany-11 committed:
feat(forget-password): add forget password screen and update routes

fix(test): remove counter test logic and keep only the main app setup

Ahmed-Abdelghany-11 committed:
feat(auth): implement JSON serialization for auth DTOs

The implementation is based on the use of  `JsonSerializableGenerator` to make it possible to serialize the data into a JSON format for the following DTOs :
 - ForgetPasswordRequestDto
 - ForgetPasswordResponseDto
 - ResetPasswordRequestDto
 - ResetPasswordResponseDto
 - VerifyResetCodeRequestDto
 - VerifyResetCodeResponseDto
Remove default test widget

Ahmed-Abdelghany-11 committed:
feat(auth): implement JSON serialization for auth DTOs

The implementation is based on the use of  `JsonSerializableGenerator` to make it possible to serialize the data into a JSON format for the following DTOs :
 - ForgetPasswordRequestDto
 - ForgetPasswordResponseDto
 - ResetPasswordRequestDto
 - ResetPasswordResponseDto
 - VerifyResetCodeRequestDto
 - VerifyResetCodeResponseDto
Remove default test widget

ali72-20 reviewed (approved):
great work keep going ahmed

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/email_verification_screen.dart:
```diff
@@ -0,0 +1,73 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final email = ModalRoute.of(context)?.settings.arguments as String;
```
good but recieve from constructor is better 

and make provider for it is more very better

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/reset_password_screen.dart:
```diff
@@ -0,0 +1,68 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final email = ModalRoute.of(context)?.settings.arguments as String;
```
good but recieve from constructor is better

and make provider for it is more very better

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,51 @@
+            autovalidateMode: AutovalidateMode.onUserInteraction,
+          ),
+          const SizedBox(height: 48),
+          SizedBox(
+            height: 48.h,
```
don't use that

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/forget_password_form.dart:
```diff
@@ -0,0 +1,51 @@
+            height: 48.h,
+            child: ElevatedButton(
+              onPressed: () {
+                if (forgetPasswordCubit.formKey.currentState?.validate() ??
+                    false) {
+                  forgetPasswordCubit.doIntent(ForgetPasswordRequestAction());
```
check on that on cubit and use valueListenable builder for to listen to it or use bloc to rebuild state for it based on formkey

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/resend_code.dart:
```diff
@@ -0,0 +1,89 @@
+class _ResendCodeState extends State<ResendCode> {
+  bool canResend = true;
+  int countdown = 0;
+  Timer? _timer;
+
+  @override
+  void initState() {
+    super.initState();
+    startCountdown();
+  }
+
+  void startCountdown() {
+    setState(() {
+      canResend = false;
+      countdown = 30;
+    });
+    _timer?.cancel();
+    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
+      setState(() {
+        countdown--;
+        if (countdown <= 0) {
+          canResend = true;
+          timer.cancel();
+        }
+      });
+    });
+  }
+
+  void onResend() {
+    context.read<EmailVerificationCubit>().doIntent(
+      ResendEmailVerificationRequestAction(widget.email),
+    );
+    startCountdown();
+  }
+
+  @override
+  void dispose() {
+    _timer?.cancel();
+    super.dispose();
+  }
```
make logic in cubit is more better

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/resend_code.dart:
```diff
@@ -0,0 +1,89 @@
+                    style: TextStyle(
+                      decoration: TextDecoration.underline,
+                      color: AppColors.pink,
+                      fontSize: 14.sp,
```
don't make sp

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/resend_code.dart:
```diff
@@ -0,0 +1,89 @@
+                  )
+                  : Text(
+                    '${LocaleKeys.ResendIn.tr()} ${countdown}s',
+                    style: TextStyle(color: AppColors.gray, fontSize: 14.sp),
```
don't make sp

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,74 @@
+                autovalidateMode: AutovalidateMode.onUserInteraction,
+              ),
+              const SizedBox(height: 48),
+              SizedBox(
+                height: 48.h,
```
don't use this way

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,74 @@
+                    ),
+                autovalidateMode: AutovalidateMode.onUserInteraction,
+              ),
+              SizedBox(height: 24.h),
```
don't use this way

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,94 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:flutter_screenutil/flutter_screenutil.dart';
+import 'package:pinput/pinput.dart';
+import 'dart:ui' as ui;
```
use dart material instead of that is more better

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,94 @@
+                forceErrorState: (state.baseState is BaseErrorState),
+                errorText: LocaleKeys.InvalidCode.tr(),
+                defaultPinTheme: PinTheme(
+                  width: 74.w,
+                  height: 68.h,
```
don't use this way

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,94 @@
+                  ),
+                ),
+                errorPinTheme: PinTheme(
+                  width: 74.w,
+                  height: 68.h,
```
don't use this way

ahmedelazab1220 commented on lib/features/forget_password/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,94 @@
+                  ),
+                  decoration: BoxDecoration(
+                    color: AppColors.white[AppColors.colorCode60],
+                    borderRadius: BorderRadius.circular(10).r,
```
don't use this way

ahmedelazab1220 commented on lib/features/forget_password/presentation/view_model/email_verification/email_verification_state.dart:
```diff
@@ -0,0 +1,33 @@
+
+sealed class EmailVerificationAction {}
+
+class EmailVerificationRequestAction extends EmailVerificationAction {}
+
+class ResendEmailVerificationRequestAction extends EmailVerificationAction {
+  final String email;
+
+  ResendEmailVerificationRequestAction(this.email);
+}
```
make this classes final is more very better

ahmedelazab1220 commented on lib/features/forget_password/presentation/view_model/forget_password_cubit/forget_password_state.dart:
```diff
@@ -1,10 +1,19 @@
-final class ForgetPasswordInitial extends ForgetPasswordState {}
+sealed class ForgetPasswordAction {}
+
+class ForgetPasswordRequestAction extends ForgetPasswordAction {}
```
make this class final is more better

ahmedelazab1220 commented on lib/features/forget_password/presentation/view_model/reset_password/reset_password_state.dart:
```diff
@@ -0,0 +1,24 @@
+
+sealed class ResetPasswordAction {}
+
+class ResetPasswordRequestAction extends ResetPasswordAction {
+  final String email;
+
+  ResetPasswordRequestAction(this.email);
+}
```
make this class final is more better

ahmedelazab1220 commented:
very very great work ya Ahmed 👏

Ahmed-Abdelghany-11 committed:
refactor(forget_password): improve email verification and reset password flows

- Add email as parameter in `ResetPasswordScreen` and `EmailVerificationScreen`
- Remove email from modal route settings in `ResetPasswordScreen` and `EmailVerificationScreen`
- Update `ResetPasswordForm` to use new email parameter
- Update `VerificationCodeInput` to show error when code is invalid
- Update `ResendCode` to manage resend logic using EmailVerificationCubit state and timer
- Refactor `ResendCode` to be stateless
- Add a new  `ForgetPasswordRequestAction` class that inherit from `ForgetPasswordAction`.
- Add a new  `ResetPasswordRequestAction` class that inherit from `ResetPasswordAction`.
- Add a new  `ResendEmailVerificationRequestAction` class that inherit from `EmailVerificationAction`.
- Implement the _startCountdown function in `EmailVerificationCubit`
- Improve all the `EmailVerificationState` class.
- Remove `screenUtil` package
- Remove unused widget test file

EslamMagdy12 committed:
Merge remote-tracking branch 'origin/developement' into feature/ta-1-forget-password

# Conflicts:
#	assets/translations/ar.json
#	assets/translations/en.json
#	lib/core/utils/di/di.config.dart
#	lib/core/utils/l10n/locale_keys.g.dart
#	lib/core/utils/routes/app_routes.dart
#	lib/core/utils/validator/validator.dart
#	lib/data/auth/api/auth_retrofit_client.dart
#	lib/data/auth/api/auth_retrofit_client.g.dart
#	lib/data/auth/data_source/contract/auth_remote_data_source.dart
#	lib/data/auth/data_source/remote/auth_remote_data_source_impl.dart
#	lib/data/auth/repo_impl/auth_repo_impl.dart
#	lib/domain/auth/repo/auth_repo.dart
#	lib/main.dart
#	test/widget_test.dart

EslamMagdy12 committed:
refactor(auth): merged dev into forget password, and ensure no issues in dart analysis.

EslamMagdy12 committed:
build: removed default widget test file

ahmedelazab1220 committed:
Merge branch 'developement' of github.com:ahmedelazab1220/tracking_flower_app into feature/ta-1-forget-password

ahmedelazab1220 committed:
feat(forget-password): refactor and enhance email verification flow with new cubit and state management

ahmedelazab1220 committed:
refactor(di.config.dart): update import paths and improve formatting for better readability

ahmedelazab1220 committed:
refactor(di.config.dart): streamline dependency registration and improve code readability

ahmedelazab1220 committed:
fix(navigation): replace pushNamed with pushReplacementNamed for email verification and reset password flows
